yarn-project/merkle-tree/src/standard_tree/unbalanced_tree.ts

```diff
@@ -0,0 +1,235 @@
+  }
+
+  /**
+   * Returns the index of a leaf given its value, or undefined if no leaf with that value is found.
+   * @param leaf - The leaf value to look for.
+   * @returns The index of the first leaf found with a given value (undefined if not found).
+   * @remark This is NOT the index as inserted, but the index which will be used to calculate path structure.
+   */
+  public findLeafIndex(value: T): bigint | undefined {
+    const key = this.valueCache[serializeToBuffer(value).toString('hex')];
+    const [, , index] = key.split(':');
+    return BigInt(index);
+  }
+
+  /**
+   * Returns the first index containing a leaf value after `startIndex`.
+   * @param value - The leaf value to look for.
+   * @param startIndex - The index to start searching from.
+   * @returns The index of the first leaf found with a given value (undefined if not found).
+   * @remark This is not really used for a wonky tree, but required to implement MerkleTree.
+   */
+  public findLeafIndexAfter(value: T, startIndex: bigint): bigint | undefined {
+    const index = this.findLeafIndex(value);
+    if (!index || index < startIndex) {
+      return undefined;
+    }
+    return index;
+  }
+
+  /**
+   * Returns the node at the given level and index
+   * @param level - The level of the element (root is at level 0).
+   * @param index - The index of the element.
+   * @returns Leaf or node value, or undefined.
+   */
+  public getNode(level: number, index: bigint): Buffer | undefined {
+    if (level < 0 || level > this.maxDepth) {
+      throw Error('Invalid level: ' + level);
+    }
+
+    if (index < 0 || index >= this.maxIndex) {
+      throw Error('Invalid index: ' + index);
+    }
+
+    return this.cache[indexToKeyHash(this.name, level, index)];
+  }
+
+  /**
+   * Returns a sibling path for the element at the given index.
+   * @param value - The value of the element.
+   * @returns A sibling path for the element.
+   * Note: The sibling path is an array of sibling hashes, with the lowest hash (leaf hash) first, and the highest hash last.
+   */
+  public getSiblingPath<N extends number>(value: bigint): Promise<SiblingPath<N>> {
+    const path: Buffer[] = [];
+    const [, depth, _index] = this.valueCache[serializeToBuffer(value).toString('hex')].split(':');
+    let level = parseInt(depth, 10);
+    let index = BigInt(_index);
+    while (level > 0) {
+      const isRight = index & 0x01n;
+      const key = indexToKeyHash(this.name, level, isRight ? index - 1n : index + 1n);
+      const sibling = this.cache[key];
+      path.push(sibling);
+      level -= 1;
+      index >>= 1n;
+    }
+    return Promise.resolve(new SiblingPath<N>(parseInt(depth, 10) as N, path));
+  }
+
+  /**
+   * Appends the given leaves to the tree.
+   * @param leaves - The leaves to append.
+   * @returns Empty promise.
+   */
+  public appendLeaves(leaves: T[]): Promise<void> {
+    this.hasher.reset();
+    if (this.size + BigInt(leaves.length) - 1n > this.maxIndex) {
```

Comment:
Don't think it will work when called repeatedly? But we don't really need to call this more than once! So maybe throw an error if this.size is not zero.

Reply:
Good point, I'd missed that! Added an error